src/ring_theory/subsemiring/basic.lean

```diff
@@ -869,16 +869,20 @@ variables [set_like σR R] [set_like σS S] [subsemiring_class σR R] [subsemiri
 open subsemiring
 
 /-- Restriction of a ring homomorphism to a subsemiring of the domain. -/
-def restrict (f : R →+* S) (s : σR) : s →+* S := f.comp $ subsemiring_class.subtype s
+def dom_restrict (f : R →+* S) (s : σR) : s →+* S := f.comp $ subsemiring_class.subtype s
 
-@[simp] lemma restrict_apply (f : R →+* S) {s : σR} (x : s) : f.restrict s x = f x := rfl
+@[simp] lemma restrict_apply (f : R →+* S) {s : σR} (x : s) : f.dom_restrict s x = f x := rfl
 
 /-- Restriction of a ring homomorphism to a subsemiring of the codomain. -/
 def cod_restrict (f : R →+* S) (s : σS) (h : ∀ x, f x ∈ s) : R →+* s :=
 { to_fun := λ n, ⟨f n, h n⟩,
   .. (f : R →* S).cod_restrict s h,
   .. (f : R →+ S).cod_restrict s h }
 
+/-- The ring homomorphism from the preimage of `s` to `s`. -/
+def restrict (f : R →+* S) (s' : subsemiring R) (s : subsemiring S) (h : s' ≤ s.comap f) :
+  s' →+* s := (f.dom_restrict s').cod_restrict s (λ x, h x.2)
+
```

Comment:
Can you please add a simp lemma `(f.restrict s' s h x : S) = f x`, and also
`(f.restrict s' s h).comp (subsemiring_class.subtype s') = subsemiring_class.subtype s`?
Thanks!

Reply:
OK, how should `(f.restrict s' s h x : S) = f x` be called? Also, `(f.restrict s' s h).comp (subsemiring_class.subtype s') = subsemiring_class.subtype s` gives me errors - how should I fix them? Thank you